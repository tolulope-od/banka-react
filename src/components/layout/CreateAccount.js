import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import Modal from '../common/Modal';
import Card from '../common/Card';
import Input from '../common/Input';
import ErrorMessage from '../common/ErrorMessage';
import createBankAccount from '../../redux/actions/accountActions';
import isEmpty from '../../utils/isEmpty';

import loadingGIF from '../../images/loading.gif';

export class CreateAccount extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isOpen: false,
      type: null,
      account: {},
      errors: {}
    };
    this.onChange = this.onChange.bind(this);
    this.onClick = this.onClick.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.modalDisplay = this.modalDisplay.bind(this);
  }

  onChange(e) {
    this.setState({ type: e.target.value });
  }

  onClick(e) {
    if (
      e.target.id === 'close-button' ||
      e.target.id === 'myModal' ||
      e.target.id === 'submit-btn'
    ) {
      this.setState({ isOpen: false });
    }
  }

  onSubmit(e) {
    e.preventDefault();
    this.setState({ error: '' });
    const { type } = this.state;
    return this.props.createBankAccount(type);
  }

  componentDidUpdate(oldProps) {
    const newProps = this.props;
    if (oldProps.account !== newProps.account) {
      this.setState({ account: newProps.account.account });
    }

    if (oldProps.errors !== newProps.errors) {
      this.setState({ errors: newProps.errors, isOpen: true });
    }
  }

  modalDisplay() {
    const { isOpen } = this.state;
    const { errors } = this.state;
    const modalSuccess = (
      <div>
        {this.props.account.accountLoading ? (
          <img src={loadingGIF} alt="loading" width="200" />
        ) : (
          <div>
            {' '}
            <img
              src="https://res.cloudinary.com/tolulope-od/image/upload/v1563465309/checkmark_cantmu.gif"
              width="200"
            />
            <h2>Account Created Successfully</h2>
            <Input
              type="submit"
              className="accnt-create-btn"
              name="addAccount"
              value="OK"
              id="submit-btn"
            />{' '}
          </div>
        )}
      </div>
    );
    const modal = isOpen ? (
      <Modal
        onClick={this.onClick}
        modalContent={
          <center>
            {!isEmpty(errors.error) ? <ErrorMessage errors={errors} /> : modalSuccess}
          </center>
        }
        className="modal-open"
      />
    ) : null;

    return modal;
  }

  render() {
    const { user } = this.props.auth;
    const modal = this.modalDisplay();
    const accountForm = (
      <form onSubmit={this.onSubmit} className="add-account">
        First name:
        <Input
          type="text"
          name="firstName"
          value={user.firstName}
          id="account-first-name"
          disabled={true}
        />
        <br />
        Last name:
        <Input
          type="text"
          name="lastName"
          value={user.lastName}
          id="account-last-name"
          disabled={true}
        />
        <br />
        <select
          name="accountType"
          className="account-type"
          id="account-type"
          onChange={this.onChange}
        >
          <option value="" className="default-slct">
            -- Select an account type
          </option>
          <option value="current" className="current-acct">
            Current
          </option>
          <option value="savings" className="savings-acct">
            Savings
          </option>
        </select>
        <Input
          type="submit"
          className="accnt-create-btn"
          name="addAccount"
          value="Submit"
          id="submit-btn"
        />
      </form>
    );
    return (
      <div>
        <div className="shortcut-btns">
          <Link to="/dashboard">
            <i className="fas fa-chevron-left"></i> Dashboard
          </Link>
        </div>
        <div className="main-dashboard">
          <div className="account-stats">
            <Card cardContent={accountForm} />
          </div>
        </div>
        {modal}
      </div>
    );
  }
}

CreateAccount.propTypes = {
  auth: PropTypes.object.isRequired,
  account: PropTypes.object,
  createBankAccount: PropTypes.func.isRequired,
  errors: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
  auth: state.auth,
  errors: state.errors,
  account: state.account
});

export default connect(
  mapStateToProps,
  { createBankAccount }
)(CreateAccount);
