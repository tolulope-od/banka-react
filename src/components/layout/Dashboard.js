import React, { Fragment, Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

import Card from '../common/Card';
import Modal from '../common/Modal';
import ErrorMessage from '../common/ErrorMessage';
import { fetchBankAccounts } from '../../redux/actions/accountActions';
import loadingGIF from '../../images/loading.gif';
import isEmpty from '../../utils/isEmpty';
import formatNumber from '../../utils/formatNumber';

export class Dashboard extends Component {
  constructor(props) {
    super(props);
    this.state = {
      accounts: [],
      isOpen: false,
      errors: {}
    };
    this.onClick = this.onClick.bind(this);
  }

  async componentDidMount() {
    try {
      const accounts = await this.props.fetchBankAccounts();
      this.setState({ accounts });
    } catch (e) {
      this.setState({ errors: { error: 'Something unusual happened, please try again' } });
    }
  }

  onClick(e) {
    if (
      e.target.id === 'close-button' ||
      e.target.id === 'myModal' ||
      e.target.id === 'submit-btn'
    ) {
      this.setState({ isOpen: false });
    }
  }

  render() {
    const { isOpen, accounts } = this.state;
    let cardContent;
    return (
      <Fragment>
        {this.props.auth.user.type === 'client' ? (
          <div className="shortcut-btns">
            <Link to="/createaccount">+ Add Account</Link>
          </div>
        ) : null}
        {this.props.account.accountLoading ? (
          <Modal
            modalContent={
              <center>
                <img src={loadingGIF} alt="loading" />
              </center>
            }
            className={isOpen ? 'modal-open' : 'modal'}
            onClick={this.onClick}
          />
        ) : (
          <div className="main-dashboard">
            <div className="account-stats" id="accnt-stats">
              {accounts.length > 0 ? (
                accounts.map(account => {
                  cardContent = (
                    <Fragment>
                      {account.status === 'active' ? (
                        <p className="badge-active" id="accnt-status">
                          {account.status}
                        </p>
                      ) : (
                        <p className="badge-inactive" id="accnt-status">
                          {account.status}
                        </p>
                      )}
                      <p className="balance-text">Balance</p>
                      <h3 className="account-balance" id="accnt-balance">
                        &#x20A6; {formatNumber(parseFloat(account.balance).toFixed(2))}
                      </h3>
                      <p className="balance-text">Account Number</p>
                      <h3 className="account-balance" id="accnt-number">
                        {account.accountnumber}
                      </h3>
                      <p className="balance-text">Account Type</p>
                      <h3 className="account-balance" id="accnt-type">
                        {account.type}
                      </h3>
                      <a
                        href="accountinfo.html"
                        className="accnt-info-btn"
                        data-account-number={account.accountnumber}
                      >
                        View Details
                      </a>
                    </Fragment>
                  );
                  return <Card key={account.id} cardContent={cardContent} />;
                })
              ) : (
                <h1>You Have No Accounts At This Time....</h1>
              )}
            </div>
          </div>
        )}
        {!isEmpty(this.props.errors.error) ? (
          <Modal
            className={isOpen ? 'modal-open' : 'modal'}
            modalContent={<ErrorMessage errors={{ error: 'An error occured, please try again' }} />}
            onClick={this.onClick}
          />
        ) : null}
      </Fragment>
    );
  }
}

Dashboard.propTypes = {
  auth: PropTypes.object,
  account: PropTypes.object.isRequired,
  fetchBankAccounts: PropTypes.func.isRequired,
  errors: PropTypes.object
};

const mapStateToProps = state => ({
  auth: state.auth,
  errors: state.errors,
  account: state.account
});

export default connect(
  mapStateToProps,
  { fetchBankAccounts }
)(Dashboard);
