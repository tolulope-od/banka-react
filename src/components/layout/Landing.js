import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import WelcomeMessage from '../common/WelcomeMessage';

import moneyProtected from '../../images/money-protected.png';
import saveMoney from '../../images/save-money.png';
import openAccount from '../../images/open-account.png';
import financeImage from '../../images/financeImage.png';

class Landing extends Component {
  render() {
    return (
      <div>
        <div className="content-container">
          <div className="main-content">
            <div className="row">
              <WelcomeMessage
                messageHead="Banking, done right"
                messageBody="Banka is the future of bank accounts management. We employ cutting edge technology
            to improve all facets of your financial life"
                optionalLink={
                  <Link to="/signup" className="auth-btn-side">
                    Get Started
                  </Link>
                }
              />
              <div>
                <img src={financeImage} alt="banking" className="illustration" />
              </div>
            </div>
          </div>
        </div>
        <h4 className="center-text section-head">Your financial partner...</h4>
        <div className="item-content-container description-content">
          <div className="column">
            <img src={openAccount} alt="Open an Account" className="description-img" />
            <p className="center-text description-txt">
              Sign up with us and create a savings or current account
            </p>
          </div>
          <div className="column">
            <img src={saveMoney} alt="Deposit money" className="description-img" />
            <p className="center-text description-txt">
              Proceed to make a deposit at any of our local branches
            </p>
          </div>
          <div className="column">
            <img src={moneyProtected} alt="Your money stays safe" className="description-img" />
            <p className="center-text description-txt">
              Your money stays secure with us against theft or fraud
            </p>
          </div>
        </div>
        <div className="item-content-container action-area">
          <Link to="/signup" className="auth-btn">
            Get Started
          </Link>
          <Link to="/login" className="auth-btn">
            Login
          </Link>
        </div>
      </div>
    );
  }
}

export default Landing;
