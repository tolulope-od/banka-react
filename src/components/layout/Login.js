import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import WelcomeMessage from '../common/WelcomeMessage';
import Input from '../common/Input';
import Loading from '../common/Loading';
import { login } from '../../redux/actions/authActions';

export class Login extends Component {
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      password: '',
      errors: {}
    };
    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.renderLoading = this.renderLoading.bind(this);
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    if (nextProps.errors !== prevState.errors) {
      return { errors: nextProps.errors };
    }

    if (nextProps.auth.isAuthenticated) {
      nextProps.history.push('/dashboard');
      return null;
    }
    return null;
  }

  componentDidMount() {
    if (this.props.auth.isAuthenticated) {
      this.props.history.push('/dashboard');
    }
  }

  onChange(e) {
    return this.setState({ [e.target.name]: e.target.value });
  }

  onSubmit(e) {
    e.preventDefault();
    const { email, password } = this.state;
    const userData = { email, password };
    this.props.login(userData);
  }

  renderLoading() {
    const { loading } = this.props.auth;
    const loadingGIF = loading ? <Loading /> : null;
    const loginBtn = loading ? (
      <Input
        type="submit"
        name="signUp"
        id="login-btn"
        value="LOADING...."
        className="btn-active"
        disabled={true}
      />
    ) : (
      <Input
        type="submit"
        name="signUp"
        id="login-btn"
        value="LOGIN"
        className="button-active"
        disabled={false}
      />
    );

    return { loadingGIF, loginBtn };
  }

  render() {
    const { errors } = this.state;
    const { loadingGIF, loginBtn } = this.renderLoading();
    return (
      <div className="content-container body-main">
        <div className="main-content">
          <div className="row">
            <WelcomeMessage
              messageHead="Banking, done right"
              messageBody="Welcome back, we missed you"
            />
            <div>
              <div className="form-card">
                <h1 className="signup-login-text">Log In</h1>
                <div className="form-body">
                  <form className="sign-up-form" onSubmit={this.onSubmit}>
                    <br />
                    <Input
                      type="email"
                      name="email"
                      placeholder="Email"
                      id="user-email"
                      value={this.state.email}
                      onChange={this.onChange}
                      required={true}
                    />
                    <br />
                    <Input
                      type="password"
                      name="password"
                      placeholder="Enter Password"
                      id="user-password"
                      value={this.state.password}
                      onChange={this.onChange}
                      required={true}
                    />
                    <br />
                    <div className="input-field-button">{loginBtn}</div>
                    <br />
                    <div className="errors">
                      <p className="auth-error-text" id="login-error">
                        {errors.error}
                      </p>
                      {loadingGIF}
                    </div>
                    <br />
                    <div className="forgot-password">
                      <Link to="/forgotPassword">Forgot Password?</Link>
                    </div>
                    <div className="signup-login-link">
                      <p>
                        Don&apos;t have an account? <Link to="/signup">Register</Link>
                      </p>
                    </div>
                  </form>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

Login.propTypes = {
  login: PropTypes.func.isRequired,
  auth: PropTypes.object.isRequired,
  errors: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
  auth: state.auth,
  errors: state.errors
});

export default connect(
  mapStateToProps,
  { login }
)(Login);
