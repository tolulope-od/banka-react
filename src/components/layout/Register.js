import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import WelcomeMessage from '../common/WelcomeMessage';
import Input from '../common/Input';
import Loading from '../common/Loading';
import { signUp } from '../../redux/actions/authActions';

export class Register extends Component {
  constructor(props) {
    super(props);
    this.state = {
      firstName: '',
      lastName: '',
      email: '',
      password: '',
      password2: '',
      errors: {},
      passwordMatchError: {}
    };
    this.onChange = this.onChange.bind(this);
    this.renderLoading = this.renderLoading.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    if (nextProps.errors !== prevState.errors) {
      return { errors: nextProps.errors };
    }
    return null;
  }

  onChange(e) {
    return this.setState({ [e.target.name]: e.target.value });
  }

  onSubmit(e) {
    e.preventDefault();
    const { firstName, lastName, email, password, password2 } = this.state;
    this.setState({ passwordMatchError: {} });
    if (password !== password2) {
      return this.setState({ passwordMatchError: { status: 401, error: 'Passwords must match' } });
    }
    const userData = { firstName, lastName, email, password };
    return this.props.signUp(userData, this.props.history);
  }

  renderLoading() {
    const { loading } = this.props.auth;
    const loadingGIF = loading ? <Loading /> : null;
    const signUpBtn = loading ? (
      <Input
        type="submit"
        name="signUp"
        id="login-btn"
        value="LOADING...."
        className="btn-active"
        disabled={true}
      />
    ) : (
      <Input
        type="submit"
        name="signUp"
        id="login-btn"
        value="SIGN UP"
        className="button-active"
        disabled={false}
      />
    );

    return { loadingGIF, signUpBtn };
  }

  render() {
    const { errors, passwordMatchError } = this.state;
    const { loadingGIF, signUpBtn } = this.renderLoading();
    return (
      <div className="content-container body-main">
        <div className="main-content">
          <div className="row">
            <WelcomeMessage
              messageHead="Banking, done right"
              messageBody="Sign up and get started"
            />
            <div>
              <div className="form-card">
                <h1 className="signup-login-text">Sign up</h1>
                <div className="form-body">
                  <form className="sign-up-form" onSubmit={this.onSubmit}>
                    <br />
                    <Input
                      type="text"
                      name="firstName"
                      placeholder="First Name"
                      id="first-name"
                      value={this.state.firstName}
                      onChange={this.onChange}
                      required={true}
                    />
                    <br />
                    <Input
                      type="text"
                      name="lastName"
                      placeholder="Last Name"
                      id="last-name"
                      value={this.state.lastName}
                      onChange={this.onChange}
                      required={true}
                    />
                    <br />
                    <Input
                      type="email"
                      name="email"
                      placeholder="Email"
                      id="user-signup-email"
                      value={this.state.email}
                      onChange={this.onChange}
                      required={true}
                    />
                    <br />
                    <Input
                      type="password"
                      name="password"
                      placeholder="Enter Password"
                      id="user-signup-password"
                      value={this.state.password}
                      onChange={this.onChange}
                      required={true}
                    />
                    <br />
                    <Input
                      type="password"
                      name="password2"
                      placeholder="Confirm Password"
                      id="confirm-password"
                      value={this.state.password2}
                      onChange={this.onChange}
                      required={true}
                    />
                    <br />
                    <div className="input-field-button">{signUpBtn}</div>
                    <br />
                    <div className="errors">
                      <p className="auth-error-text" id="login-error">
                        {passwordMatchError.error}
                        {errors.error}
                      </p>
                      {loadingGIF}
                    </div>
                    <br />
                    <div className="forgot-password">
                      <Link to="/forgotPassword">Forgot Password?</Link>
                    </div>
                    <div className="signup-login-link">
                      <p>
                        Have an account? <Link to="/login">Login</Link>
                      </p>
                    </div>
                  </form>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

Register.propTypes = {
  signUp: PropTypes.func.isRequired,
  auth: PropTypes.object.isRequired,
  errors: PropTypes.object.isRequired,
  history: PropTypes.object
};

const mapStateToProps = state => ({
  auth: state.auth,
  errors: state.errors
});

export default connect(
  mapStateToProps,
  { signUp }
)(Register);
