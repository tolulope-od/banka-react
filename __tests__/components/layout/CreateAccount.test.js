import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { CreateAccount } from '../../../src/components/layout/CreateAccount';

const shallowSetup = () => {
  const props = {
    auth: {
      isAuthenticated: true,
      user: {
        firstName: 'Test',
        lastName: 'User'
      }
    },
    account: {
      accountLoading: false,
      account: {}
    },
    createBankAccount: jest.fn(),
    errors: {}
  };

  const enzymeWrapper = shallow(<CreateAccount {...props} />);

  return { props, enzymeWrapper };
};

describe('<CreateAccount /> Component', () => {
  const { enzymeWrapper } = shallowSetup();
  const component = enzymeWrapper.instance();

  it('renders the create account page', () => {
    expect(enzymeWrapper.find('.shortcut-btns').length).toBe(1);
    expect(enzymeWrapper.find('Link').length).toBe(1);
    expect(enzymeWrapper.find('Card').length).toBe(1);
  });

  it('should set the state when the component is updated', () => {
    enzymeWrapper.setProps({ errors: { error: 'Something happened' } });
    expect(component.state.errors.error).toEqual('Something happened');
    expect(component.state.isOpen).toEqual(true);
  });

  it('should change the account type in local state', () => {
    const e = {
      target: {
        name: 'accountType',
        value: 'current'
      }
    };
    component.onChange(e);
    expect(component.state.type).toEqual('current');
  });

  it('should create an account when the button is clicked', () => {
    const createAccount = jest.spyOn(component, 'onSubmit');
    component.onSubmit({ preventDefault: () => 1 });
    expect(createAccount).toBeCalled();
  });

  it('should set the state when an account is created', () => {
    const newAccount = {
      id: 35,
      accountnumber: '6407771369',
      owner: 5,
      ownername: 'Joy Machido',
      type: 'savings',
      status: 'active',
      balance: '0.0',
      owneremail: 'joy.machido@gmail.com',
      createdon: '2019-08-20T13:22:41.935Z'
    };
    enzymeWrapper.setProps({
      account: {
        account: newAccount
      }
    });
    expect(component.state.account).toEqual(newAccount);
  });

  it('should call the onClick function', () => {
    const clickAction = jest.spyOn(component, 'onClick');
    component.onClick({ target: { id: 'modal' } });
    expect(clickAction).toBeCalled();
  });

  it('should close the modal when the close button is clicked', () => {
    const e = {
      target: {
        id: 'close-button'
      }
    };
    component.onClick(e);
    expect(component.state.isOpen).toEqual(false);
  });

  it('should close the modal when the submit button is clicked', () => {
    const e = {
      target: {
        id: 'submit-btn'
      }
    };
    component.onClick(e);
    expect(component.state.isOpen).toEqual(false);
  });

  it('should close the modal when any area outside the modal is clicked', () => {
    const e = {
      target: {
        id: 'myModal'
      }
    };
    component.onClick(e);
    expect(component.state.isOpen).toEqual(false);
  });

  it('Should change the content in the modal if the account information is loading', () => {
    const displayedModal = jest.spyOn(component, 'modalDisplay');
    const modal = component.modalDisplay();

    expect(toJson(modal)).toMatchSnapshot();
    expect(displayedModal).toBeCalled();
  });
});
