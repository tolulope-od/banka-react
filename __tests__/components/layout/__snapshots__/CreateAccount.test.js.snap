// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateAccount /> Component Should change the content in the modal if the account information is loading 1`] = `undefined`;
